### PR TITLE
Parallel RDS Spinup

### PR DESCRIPTION
**Pull request by pchand-amplify**
This PR enables Parallel execution of RDS instances. The following changes have been made
- Use of python Threading library.
- A new class derived from Threading, that encapsulates the creation/update of a single RDS instance.
- Option to enable or disable parallel execution.
- Add `--parallel` option to `disco_rds.py update`.

### CODE REVIEW

**Comment by pchand-amplify** [comment]
Sorry about the massive code change. Most of the changes are code being moved from one class to the other.

File: disco_aws_automation/disco_rds.py [hunk 1]
```diff
@@ -170,80 +381,142 @@ def update_cluster(self, database_name):
...
+
+    def __init__(self, vpc):
+        """Initialize class"""
+        self.config_aws = read_config()
+        self.config_rds = read_config(config_file=DEFAULT_CONFIG_FILE_RDS)
+        self.client = boto3.client('rds')
+        self.vpc = vpc
+        self.disco_vpc_sg_rules = DiscoVPCSecurityGroupRules(vpc, vpc.boto3_ec2)
+        self.vpc_name = vpc.environment_name
+        if self.vpc_name not in ['staging', 'production']:
```
Comment (jconstance-amplify) [inline disco_aws_automation/disco_rds.py]: This seems unnecessarily hard-coded. Might be better to copy-paste the config logic for properly handling the `param@env` syntax than to hard code VPC names that are not necessarily universal.

Another point in favor of the eventual `DiscoConfig` class.

File: disco_aws_automation/disco_rds.py [hunk 2]
```diff
@@ -137,29 +284,93 @@ def get_instance_parameters(self, env_name, database_name):
...
+                                     'Description': 'Description',
+                                     'Source': 'engine-default',
+                                     'ApplyType': 'static',
+                                     'DataType': 'string',
+                                     'AllowedValues': 'somevalues',
+                                     'IsModifiable': True,
+                                     'MinimumEngineVersion': 'someversion',
+                                     'ApplyMethod': 'pending-reboot'}])
+
+    def recreate_db_parameter_group(self, env_name, database_name, db_parameter_group_name,
```
Comment (kelibezhani) [inline disco_aws_automation/disco_rds.py]: I do not understand why we have `create_db_parameter_group`, `modify_db_parameter_group`, and `recreate_db_parameter_group`. Can't we implement *create* and *recreate* in the same function?

Comment (kelibezhani) [inline disco_aws_automation/disco_rds.py]: This has already been tested. I do not have any significant remarks, overall looks good to me. 👍

File: disco_aws_automation/disco_rds.py (hunk 1 above)
Comment (jconstance-amplify) [inline disco_aws_automation/disco_rds.py]: Ah, I see that you didn't introduce this. I guess okay for now. But we should really fix this.

**Comment by jconstance-amplify** [comment]
👍

File: disco_aws_automation/disco_rds.py [hunk 3]
```diff
@@ -39,93 +40,239 @@
...
-    """Class for doing RDS operations on a given environment"""
+class RDS(threading.Thread):
+    """Class for spinning up a single rds instance"""
 
-    def __init__(self, vpc):
+    def __init__(self, env_name, database_identifier, rds_security_group_id, subnet_ids, domain_name):
         """Initialize class"""
+        threading.Thread.__init__(self, name=database_identifier)
+        self.client = boto3.client('rds')
+        self.vpc_name = env_name
```
Comment (jichen-amplify) [inline disco_aws_automation/disco_rds.py]: nitpick: looks like we can just derive the vpc_name from the database_identifier without having an extra argument to the method.

File: disco_aws_automation/disco_rds.py [hunk 4]
```diff
@@ -170,80 +381,142 @@ def update_cluster(self, database_name):
...
+
+        # For Sequential execution
+        if not parallel:
+            for rds in rds_list:
+                rds.update_cluster()
+            return
+
+        # Running in parallel
+        for rds in rds_list:
+            rds.start()
```
Comment (jichen-amplify) [inline disco_aws_automation/disco_rds.py]: nit: is it ok to call `.join()` right after `.start()`? We could avoid an extra for loop if yes.

**Comment by jichen-amplify** [comment]
👍

**Comment by pchand-amplify** [comment]
build

**Pull request merged**